src/network/bpf/sysctl_monitor/sysctl-monitor.bpf.c

```diff
@@ -0,0 +1,99 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "vmlinux.h"
+
+#include <bpf/bpf_helpers.h>
+
+#include "sysctl-write-event.h"
+
+struct {
+        __uint(type, BPF_MAP_TYPE_CGROUP_ARRAY);
+        __type(key, u32);
+        __type(value, u32);
+        __uint(max_entries, 1);
+} cgroup_map SEC(".maps");
+
+struct {
+        __uint(type, BPF_MAP_TYPE_RINGBUF);
+        __uint(max_entries, 256 * 1024);
+} written_sysctls SEC(".maps");
+
+static bool my_streq(const char *s1, const char *s2, size_t l) {
+        for (size_t i = 0; i < l; i++) {
+                if (s1[i] != s2[i])
+                        return false;
+                if (s1[i] == 0)
+                        return true;
+        }
+        return true;
+}
+
+struct str {
+        char *s;
+        size_t l;
+};
+
+static long cut_last(u32 i, struct str *str) {
+        char *s;
+
+        i = str->l - i - 1;
+        s = str->s + i;
+
+        /* Sanity check for the preverifier */
+        if (i >= str->l)
+                return 1;
+
+        if (*s == 0)
+                return 0;
+
+        if (*s == '\n' || *s == '\r' || *s == ' ' || *s == '\t') {
+                *s = 0;
+
+                return 0;
+        }
+
+        return 1;
+}
+
+/* Cut off trailing whitespace and newlines */
+static void chop(char *s, size_t l) {
+        struct str str = { s, l };
+
+        bpf_loop(l, cut_last, &str, 0);
+}
+
+SEC("cgroup/sysctl")
+int sysctl_monitor(struct bpf_sysctl *ctx) {
+        struct sysctl_write_event we = { };
+
+        /* Ignore events generated by us */
+        if (bpf_current_task_under_cgroup(&cgroup_map, 0))
+                return 1;
+
+        /* Allow reads */
+        if (!ctx->write)
+                return 1;
+
```

Comment:
i'd move the declaratoin here, and initialize the version right-away, i.e. insert here:

```c
struct sysctl_write_event we = { 
        .version = 1,
};
``

Reply:
I set the version later on purpose because setting the version here initializes the whole structure to zero.
This makes the BPF object 520 bytes bigger and waste CPU cycles even when we don't have to raise any event.

Reply:
hmm, but you already initialized we at time of declaration before, just to {}?

all my suggestion changes is that we allocate it later, and that we initialize the contents properly right away, instead of initialzing it to zero first, and then overriding it again.

Reply:
ah, you are not initializing it to zero anymore

Reply:
hmm, or maybe all local variables are zero initialized for security reasons anyway in bpf?

Reply:
`{}` was a leftover, because initially I was doing `{ .version = 1 }`, but then I noticed a huge increase in code size.
The only way to avoid the zero initialization is to omit curly braces at all.
The values copied in the struct are usually 10-20 bytes each, there is no need to zero fill the whole 460 bytes struct at every event.
The problem with BPF is that instructions can't never jump backwards, and this makes loops impossible (this is why there is the `bpf_loop` helper).
So zero filling a big struct is unrolled in hundreds of instructions.

In this case, this code:
```c
        struct sysctl_write_event we;
        we.version = 1;
```
leads to this object:
```
5432 Sep 10 13:23 build/src/network/bpf/sysctl_monitor/sysctl-monitor.bpf.o
```

while this:

```c
        struct sysctl_write_event we = {
                .version = 1,
        };
```
leads to this object, 960 bytes bigger:
```
6392 Sep 10 13:26 build/src/network/bpf/sysctl_monitor/sysctl-monitor.bpf.o
```